### ISSUE TITLE
Is it possible to delay trigger at end of transaction?  Add deferrable option?

### ISSUE THREAD

**Issue by rrauenza** [Opus10/django-pgtrigger#14]
Let's say I have a model... just don't question why my car's only have one wheel :)

```
class Car:
     wheel = OneToOneField('Wheel', null=True)
     requires_wheel = BooleanField()
```

I have a constraint that if requires_wheel is True, then wheel can't be None.  But this can leave dangling wheels when Cars are deleted.  So I could do this instead and get cascading deletes:

```
class Wheel:
    car = OneToOneField('Car')
```

But how do I enforce that if requires_wheel is True that the Wheel also exists?  I was thinking through a trigger, but the trigger would need to wait until the end of the transaction.  Deferrable?

**Comment by jjorissen52** [comment]
Maybe you could accomplish this using Statement level triggers?

https://django-pgtrigger.readthedocs.io/en/latest/tutorial.html#statement-level-triggers-and-transition-tables

**Comment by wesleykendall** [comment]
@rrauenza it sounds like you want two different triggers:

1. A "Before" trigger that will ensure wheel isn't null when requires_wheel is True before any database operation happens.
2. An "After delete" trigger for the "Car" model that will delete the "Wheel" model if the wheel has no more references

**Comment by wesleykendall** [comment]
@rrauenza I looked into the issue of constraint triggers just a bit more. According to https://www.postgresql.org/docs/10/sql-createtrigger.html, only constraint triggers have the ability to work at the end of a transaction.

I need to do a bit more digging on this to see how to proceed. I'm not totally sure if constraint triggers should be in their own subclass or if it might make sense to just put them in the main class since they have restrictions and options other triggers don't have

I'm also considering mysql support for this library too, so I want to take that into account before putting them in.

Have you had other uses for constraint triggers? Will get back to this issue in a week or two. Going to get through some of the others first, but let me know if you have any ideas!

**Comment by rrauenza** [comment]
I don't remember what I did ... but I worked around it or just decided I didn't need it.

**Comment by peterthomassen** [comment]
@wesleykendall I have a model `TokenDomainPolicy` with a trigger that imposes a certain consistency check across rows upon write operations; it requires certain values to be inserted in a certain order (I can explain in more detail if needed). The table has a foreign key column that relates it to `Token`, with `on_delete=models.CASCADE` in the Django model definition. If I now delete a `Token` instance, Django takes care of the cascade, and deletes the related `TokenDomainPolicy` rows in an order that violates the trigger's consistency check. If the trigger was delayed to the end of the transaction, the view would be consistent and all would be fine.

Do you see the possibility that handling this may soon be possible with `django-pgtrigger`?

**Comment by wesleykendall** [comment]
@peterthomassen I read into things more, and it looks like Postgres has constraint triggers for deferring execution to the end of a COMMIT. Can you verify that this is what would solve your use case?

For reference, I read [this article](https://www.cybertec-postgresql.com/en/triggers-to-enforce-constraints/#:~:text=By%20making%20the%20trigger%20INITIALLY,t%20see%20each%20other's%20modifications.) when I was investigating your problem.

I think this could be an easy addition to the library

**Comment by peterthomassen** [comment]
Indeed, [one of the triggers we use](https://github.com/desec-io/desec-stack/blob/1879bad0fbeaab6fa365497998ef5f197da9add2/api/desecapi/models.py#L529-L538) could be better expressed as a deferred trigger.

**Comment by wesleykendall** [comment]
@peterthomassen FYI here's how I'm planning to address this:

1. Add a `timing` attribute that defaults to `None`. If set to `InitiallyImmediate` or `InitiallyDeferrable`, we will install a `CONSTRAINT` trigger that can be deferred. I was originally going to have both a `deferrable` option and `timing` option for triggers, but I'm not sure what value this would bring. To the best of my knowledge, a non-deferrable constraint trigger has the same behavior as a non-constraint trigger, so I don't see a need to allow this second option.
2. Add a function called `pgtrigger.constraints` that will execute `SET CONSTRAINTS` to dynamically alter deferrable constraint timing in a transaction. I would put this in a context manager, but I haven't been able to find a way to reset this value back to the default. It seems like the intended use of this is to set it once within a transaction, so I'm planning to implement it that way and document its usage. I don't think you need this for your use case, but I wanted to add it for completeness

Going to get a PR together for this now. I'm likely going to do a release with a few other requested features later tonight. Will let you know when it's deployed

Also FYI - Checked out your code, and you can now declare triggers in `Meta` with the `triggers` attribute since version 2.5. Version 3 integrates pgtrigger into the migration system too

**Comment by peterthomassen** [comment]
Nice, thanks for working on all of that! <3

Regarding your proposal, it sounds good to me. But I'm not a "fluent" Postgres expert -- I can make reasonable judgments, but to do that, I need to spend considerable time re-familiarizing myself with the Postgres docs etc. (don't have it all in my head). So, while it sounds reasonable to me, it is not an expert opinion at this moment. Others who deal with Postgres on a more frequent basis than me should chime in if needed.

**Comment by wesleykendall** [comment]
You're welcome! I did the initial feature in https://github.com/Opus10/django-pgtrigger/pull/74. I added a test case to verify triggers are deferred, along with verifying that they can be ignored too.

Feature will be deployed tonight, and I'll add some notes here when it is. I'm going to try to get a few other things in the release for tonight.

**Comment by wesleykendall** [comment]
Released in version 3.4!

There is a special section [in the docs](https://django-pgtrigger.readthedocs.io/en/latest/deferrable.html) for deferrable triggers, along with an example that's similar to the original example here.

As noted before, when you upgrade past version 3, triggers are integrated with the migration system and also support `Meta.triggers` syntax now. See #67 for details on upgrading
